README
- sIBL_GUI_XSI_Server - Addon -

- Installation -
Extract and Install like an usual XSI Addon: XSI -> File - > Addon -> Install Addon.

- Usage -
sIBL_GUI_XSI_Server menu entry is stored in Model -> Get -> Primitive -> Light. "sIBL_GUI_XSI_Server Properties" launches the sIBL_GUI_XSI_Server window.

- About -
sIBL_GUI by Thomas Mansencal - 2008 - 2010
Copyright (C) 2008 - 2010 - Thomas Mansencal - thomas.mansencal@gmail.com
This software is released under terms of GNU GPL V3 license: http://www.gnu.org/licenses/
http://www.thomasmansencal.com/

CHANGES
h1=. sIBL_GUI XSI Server - 1.0.0

h2=. Change Log

h3. 1.0.0

* Initial Release of sIBL_GUI XSI Server. Notice that the Server is really stable and should'nt receive many updates.

h2=. About

sIBL_GUI by Thomas Mansencal - 2008 - 2014
Copyright (C) 2008 - 2014 - Thomas Mansencal - "thomas.mansencal@gmail.com":mailto:thomas.mansencal@gmail.com
This Software Is Released Under Terms Of GNU GPL V3 License: "http://www.gnu.org/licenses/":http://www.gnu.org/licenses/
"http://www.thomasmansencal.com/":http://www.thomasmansencal.com/

Addons/sIBL_GUI_XSI_Server/cssrc/sIBL_GUI_XSI_Server.cs
// sIBL_GUI_XSI_ServerPlugin

using System;
using System.Net;
using XSI.TCP;
using XSI.Helpers;
using Softimage.XSIOM; // XSI object model

// Defining Global Variables
public class sIBL_GUI_XSI_Server_Globals
{
    public static string Address = "127.0.0.1";
    public static int Port = 12288;
    public static int MaxCnx = 10;
}

public class XSIPlugin: Base
{
    public bool Load(PluginRegistrar in_reg)
    {
        in_reg.Author = "Softimage";
        in_reg.Name = "sIBL_GUI_XSI_ServerPlugin";
        in_reg.Major = 1;
        in_reg.Minor = 0;

        in_reg.RegisterCommand("sIBL_GUI_XSI_Server_Start", null);
        in_reg.RegisterCommand("sIBL_GUI_XSI_Server_Stop", null);
        in_reg.RegisterTimerEvent("sIBL_GUI_XSI_Server_ProcessRequestsTimerEvent", 500, 0);
        TimerEvent evTimer = (TimerEvent)GetXSI().EventInfos["sIBL_GUI_XSI_Server_ProcessRequestsTimerEvent"];
        evTimer.Mute = true;

        in_reg.RegisterMenu(siMenuAnchorPoints.siMenuTbGetLightID, "sIBL_GUI_XSI_Server_Menu", false, false);
        in_reg.RegisterProperty("sIBL_GUI_XSI_Server_Property");
        in_reg.RegisterEvent("sIBL_GUI_XSI_Server_StartupEvent", siEventID.siOnStartup);
        //RegistrationInsertionPoint - do not remove this line

        return true;
    }

    public bool Unload(PluginRegistrar in_reg)
    {
        String strPluginName = null;
        strPluginName = in_reg.Name;
        Log(strPluginName + " has been unloaded.");

        // stop server
        sIBL_GUI_XSI_Server.Stop();

        return true;
    }
}

// Timer event calback for executing client requests in XSI.
// Note: This callback is necessary in order to perform XSI operations in the
// main thread. Calling XSI operations from worker threads may put XSI in an
// undefined state.
public class sIBL_GUI_XSI_Server_ProcessRequestsTimerEvent: Base
{
    public bool OnEvent(Context in_ctxt)
    {
        ClientRequests requests = new ClientRequests();
        requests.Process();
        return false; // return false if we don't want to mute the timer
    }
}

// The XSI tcp server implementation
public class sIBL_GUI_XSI_Server
{
    public static RequestsService m_provider = null;
    public static Server m_server = null;
    public static void Start(String in_address, int in_port, int in_max)
    {
        if (m_server == null)
        {
            // Creates a TCP server to execute client requests asynchronously.
            m_provider = new RequestsService();
            m_server = new Server(m_provider, in_port, in_max);
            m_server.Start(in_address);
        }
    }
    public static void Stop()
    {
        if (m_server != null)
        {
            m_server.Stop();
            m_server = null;
        }
    }
}

// sIBL_GUI_XSI_Server_Start command
public class sIBL_GUI_XSI_Server_Start: Base
{
    public bool Execute(Context in_ctxt)
    {
        TimerEvent evTimer = (TimerEvent)GetXSI().EventInfos["sIBL_GUI_XSI_Server_ProcessRequestsTimerEvent"];

        try
        {
            //Starting With Default Values
            sIBL_GUI_XSI_Server.Start(sIBL_GUI_XSI_Server_Globals.Address,
                    sIBL_GUI_XSI_Server_Globals.Port,
                    sIBL_GUI_XSI_Server_Globals.MaxCnx);
            evTimer.Mute = false;
        }
        catch
        {
            sIBL_GUI_XSI_Server.Stop();
            evTimer.Mute = true;
        }
        return true;
    }
}

// sIBL_GUI_XSI_Server_Stop command
public class sIBL_GUI_XSI_Server_Stop: Base
{
    public bool Execute(Context in_ctxt)
    {
        Info("sIBL_GUI_XSI_Server | Server Stopped!");
        sIBL_GUI_XSI_Server.Stop();
        TimerEvent evTimer = (TimerEvent)GetXSI().EventInfos["sIBL_GUI_XSI_Server_ProcessRequestsTimerEvent"];
        evTimer.Mute = true;
        return true;
    }
}

// sIBL_GUI_XSI_Server menu
public class sIBL_GUI_XSI_Server_Menu: Base
{
    public bool Init(Context in_ctxt)
    {
        Menu oMenu = null;
        oMenu = (Menu)in_ctxt.Source;
        oMenu.AddCallbackItem("sIBL_GUI_XSI_Server Properties", "OnOpensIBL_GUI_XSI_ServerPPG");
        return true;
    }

    public bool OnOpensIBL_GUI_XSI_ServerPPG(Context in_ctxt)
    {
        try
        {
            Array inspectobjArgs = Array.CreateInstance(typeof(Object), 5);
            inspectobjArgs.SetValue(sIBL_GUI_XSI_Server_Property.PSet(GetXSI()), 0);
            GetXSI().ExecuteCommand("InspectObj", inspectobjArgs);
        }
        catch (Exception e)
        {
            Error(e.ToString());
        }
        return true;
    }
}

//
// sIBL_GUI_XSI_Server_Property: Provides the UI for managing the sIBL_GUI_XSI_Server
//

public class sIBL_GUI_XSI_Server_Property: Base
{
    public bool Define(Context in_ctxt)
    {
        try
        {
            CustomProperty oCustomProperty = (CustomProperty)in_ctxt.Source;
            oCustomProperty.AddParameter2("Address", siVariantType.siString, sIBL_GUI_XSI_Server_Globals.Address, null, null, null, null, siParamClassification.siClassifUnknown, (int)(siCapabilities.siPersistable | siCapabilities.siReadOnly), null, null, null);
            oCustomProperty.AddParameter2("Port", siVariantType.siInt4, sIBL_GUI_XSI_Server_Globals.Port, 10000, 65536, 10000, 65536, siParamClassification.siClassifUnknown, (int)siCapabilities.siPersistable, null, null, null);
            oCustomProperty.AddParameter2("MaxCnx", siVariantType.siInt4, sIBL_GUI_XSI_Server_Globals.MaxCnx, 1, 20, 1, 20, siParamClassification.siClassifUnknown, (int)siCapabilities.siPersistable, null, null, null);
            oCustomProperty.AddParameter2("ProcessReqTimer", siVariantType.siInt4, 500, 100, 10000, 100, 10000, siParamClassification.siClassifUnknown, (int)siCapabilities.siPersistable, null, null, null);
            oCustomProperty.AddParameter2("EnableRequests", siVariantType.siBool, true, null, null, null, null, siParamClassification.siClassifUnknown, (int)siCapabilities.siPersistable, null, null, null);
        }
        catch (Exception e)
        {
            Error(e.ToString());
        }

        return true;
    }

    public bool DefineLayout(Context in_ctxt)
    {
        try
        {
            PPGLayout oLayout = (PPGLayout)in_ctxt.Source;
            oLayout.Clear();

            oLayout.AddGroup("Server", true, 0);

            oLayout.AddGroup(null, true, 0);
            oLayout.AddRow();
            oLayout.AddButton("StartServer", "Start");
            oLayout.AddButton("StopServer", "Stop");
            oLayout.EndRow();
            oLayout.EndGroup();

            oLayout.AddGroup(null, true, 0);
            oLayout.AddItem("Address", "Address", null);
            oLayout.AddRow();
            oLayout.AddButton("GetHostAddress", "Use Host Address");
            oLayout.AddButton("GetLocalHostAddress", "Use Local Host Address");
            oLayout.EndRow();
            oLayout.EndGroup();

            oLayout.AddItem("Port", "Port", null);
            oLayout.AddItem("MaxCnx", "Maximum Connections", null);
            oLayout.EndGroup();

            oLayout.AddGroup("Requests", true, 0);
            oLayout.AddItem("EnableRequests", "Enable Processing", null);
            oLayout.AddItem("ProcessReqTimer", "Processing Interval (ms)", null);

            oLayout.AddGroup(null, true, 0);
            oLayout.AddRow();
            oLayout.AddButton("LogRequests", "Log");
            oLayout.AddButton("ClearRequests", "Clear");
            oLayout.EndRow();
            oLayout.EndGroup();

            oLayout.EndGroup();
        }
        catch (Exception e)
        {
            Error(e.ToString());
        }

        return true;
    }

    public bool PPGEvent(Context in_ctxt)
    {
        try
        {
            PPGEventContext ppgctxt = (PPGEventContext)in_ctxt;
            siPPGEventID eventID = ppgctxt.EventID;

            if (eventID == siPPGEventID.siParameterChange)
            {
                // The Source of the event is the parameter itself
                Parameter changed = (Parameter)ppgctxt.Source;
                if ("ProcessReqTimer" == changed.ScriptName)
                {
                    int interval = (int)changed.GetValue2(null);
                    TimerEvent evTimer = (TimerEvent)GetXSI().EventInfos["sIBL_GUI_XSI_Server_ProcessRequestsTimerEvent"];
                    evTimer.Reset(interval, 0);
                }
                else if ("EnableRequests" == changed.ScriptName)
                {
                    bool bValue = (bool)changed.GetValue2(null);
                    EventInfo evTimer = GetXSI().EventInfos["sIBL_GUI_XSI_Server_ProcessRequestsTimerEvent"];
                    evTimer.Mute = bValue == false;
                }

            }
            else if (eventID == siPPGEventID.siButtonClicked)
            {
                String buttonPressed = (String)ppgctxt.GetAttribute("Button");

                if (buttonPressed == "GetHostAddress")
                {
                    // Set host param with this host address
                    IPHostEntry hostEntry = Dns.GetHostEntry(Dns.GetHostName());
                    IPAddress[] ipAddresses = hostEntry.AddressList;

                    Property prop = (Property)ppgctxt.Source;
                    Parameter adr = prop.Parameters["Address"];
                    adr.PutValue2(null, ipAddresses[0].ToString());
                }
                else if (buttonPressed == "GetLocalHostAddress")
                {
                    // Set host param with this host address
                    Property prop = (Property)ppgctxt.Source;
                    Parameter adr = prop.Parameters["Address"];
                    adr.PutValue2(null, "127.0.0.1");
                }
                else if (buttonPressed == "LogRequests")
                {
                    Info("Logging all requests ...");
                    ClientRequests requests = new ClientRequests();
                    requests.Log();
                }
                else if (buttonPressed == "ClearRequests")
                {
                    Info("Removing all requests ...");
                    ClientRequests requests = new ClientRequests();
                    requests.Clear();
                }
                else if (buttonPressed == "StartServer" || buttonPressed == "StopServer")
                {
                    TimerEvent evTimer = (TimerEvent)GetXSI().EventInfos["sIBL_GUI_XSI_Server_ProcessRequestsTimerEvent"];
                    if (buttonPressed == "StartServer")
                    {
                        // Start server
                        Info("sIBL_GUI_XSI_Server | Server Started!");
                        Property prop = (Property)ppgctxt.Source;
                        sIBL_GUI_XSI_Server.Start((String)prop.Parameters["Address"].GetValue2(null),
                                    (int)prop.Parameters["Port"].GetValue2(null),
                                    (int)prop.Parameters["MaxCnx"].GetValue2(null));
                        evTimer.Mute = false;
                    }
                    else if (buttonPressed == "StopServer")
                    {
                        Info("sIBL_GUI_XSI_Server | Server Stopped!");
                        sIBL_GUI_XSI_Server.Stop();
                        evTimer.Mute = true;
                    }
                }
            }
        }
        catch (Exception e)
        {
            Error(e.ToString());
        }

        return true;
    }

    static public Property PSet(XSIApplication in_xsi)
    {
        Property prop = null;
        try
        {
            Model root = in_xsi.ActiveSceneRoot;
            prop = root.Properties["sIBL_GUI_XSI_Server_Property"];

            if (prop == null)
            {
                prop = root.AddProperty("sIBL_GUI_XSI_Server_Property", false, "sIBL_GUI_XSI_Server_Property");
            }
        }
        catch (Exception e)
        {
            in_xsi.LogMessage(e.ToString(), siSeverity.siError);
        }
        return prop;
    }

}

// sIBL_GUI_XSI_Server_StartupEvent
public class sIBL_GUI_XSI_Server_StartupEvent: Base
{
    public bool OnEvent(Context in_ctxt)
    {
        GetXSI().ExecuteCommand("sIBL_GUI_XSI_Server_Start", null);

        return true;
    }
}

Addons/sIBL_GUI_XSI_Server/Application/Plugins/tcpServer.py
"""
**tcpServer.py**

**Platform:**
	Windows, Linux.

**Description:**
	This module defines the :class:`TCPServer` and :class:`RequestHandler` classes.

**Others:**

"""

#**********************************************************************************************************************
#***	External imports.
#**********************************************************************************************************************
import SocketServer
import sys
import threading


#**********************************************************************************************************************
#***	Module attributes.
#**********************************************************************************************************************
__author__ = "Thomas Mansencal"
__copyright__ = "Copyright (C) 2008 - 2012 - Thomas Mansencal"
__license__ = "GPL V3.0 - http://www.gnu.org/licenses/"
__maintainer__ = "Thomas Mansencal"
__email__ = "thomas.mansencal@gmail.com"
__status__ = "Production"

__all__ = []

#**********************************************************************************************************************
#***	Module classes and definitions.
#**********************************************************************************************************************
class DefaultRequestHandler(SocketServer.BaseRequestHandler):

	def handle(self):
		while True:
			data = self.request.recv(1024)
			if len(data) == 0:
				break
			sys.stdout.write(data)
		return True

class TCPServer(object):

	def __init__(self, address, port, handler=DefaultRequestHandler):
		self.__address = None
		self.address = address
		self.__port = None
		self.port = port
		self.__handler = None
		self.handler = handler

		self.__server = SocketServer.TCPServer((address, port), handler)
		self.__worker = None

	#******************************************************************************************************************
	#***	Attributes properties.
	#******************************************************************************************************************
	@property
	def address(self):
		return self.__address

	@address.setter
	def address(self, value):
		if value is not None:
			assert type(value) in (str, unicode), "'{0}' attribute: '{1}' type is not 'str' or 'unicode'!".format(
			"address", value)
		self.__address = value

	@address.deleter
	def address(self):
		raise Exception("{0} | '{1}' attribute is not deletable!".format(self.__class__.__name__, "address"))

	@property
	def port(self):
		return self.__port

	@port.setter
	def port(self, value):
		if value is not None:
			assert type(value) is int, "'{0}' attribute: '{1}' type is not 'int'!".format(
			"port", value)
		self.__port = value

	@port.deleter
	def port(self):
		raise Exception("{0} | '{1}' attribute is not deletable!".format(self.__class__.__name__, "port"))

	@property
	def handler(self):
		return self.__handler

	@handler.setter
	def handler(self, value):
		if value is not None:
			assert issubclass(value, SocketServer.BaseRequestHandler), \
			"'{0}' attribute: '{1}' is not 'SocketServer.BaseRequestHandler' subclass!".format("handler", value)
		self.__handler = value

	@handler.deleter
	def handler(self):
		raise Exception("{0} | '{1}' attribute is not deletable!".format(self.__class__.__name__, "handler"))

	#******************************************************************************************************************
	#***	Class methods.
	#******************************************************************************************************************
	def start(self):
		self.__worker = threading.Thread(target=self.__server.serve_forever)
		self.__worker.setDaemon(True)
		self.__worker.start()
	
	def stop(self):
		self.__server.socket.close()
		self.__server.shutdown()


Addons/sIBL_GUI_XSI_Server/cssrc/sIBL_GUI_XSI_Server_TCP.cs
using System;
using System.Net;
using System.Net.Sockets;
using System.Threading;
using System.Text;
using System.Collections;

namespace XSI.TCP
{
	// Class to keep track of each client connected to the server, and to
	// provide the implementation for sending/receiving data to the remote
	// sIBL_GUI_XSI_Server.
	public class ConnectionState
	{
		internal Socket m_cnx;
		internal Server m_server;
		internal ServiceProvider m_provider;
		internal byte[] m_buffer;

		// The IP Address of the remote host.
		public EndPoint RemoteEndPoint
		{
			get{ return m_cnx.RemoteEndPoint; }
		}

		// Returns the number of bytes waiting to be read.
		public int AvailableData
		{
			get{ return m_cnx.Available; }
		}

		// Returns true if the socket is connected.
		public bool Connected
		{
			get{ return m_cnx.Connected; }
		}

		// Reads data from the socket and returns the number of bytes read.
		public int ReadFrom(byte[] in_buffer, int in_offset, int in_count)
		{
			try
			{
				if (m_cnx.Available > 0)
				{
					return m_cnx.Receive(	in_buffer,
											in_offset,
											in_count,
											SocketFlags.None);
				}
				else
				{
					return 0;
				}
			}
			catch
			{
				return 0;
			}
		}

		// Sends Data to the remote host.
		public bool WriteTo(byte[] in_buffer, int in_offset, int in_count)
		{
			try
			{
				m_cnx.Send(in_buffer, in_offset, in_count, SocketFlags.None);
				return true;
			}
			catch
			{
				return false;
			}
		}

		// Terminates the connection with the remote host.
		public void EndConnection()
		{
			if(m_cnx != null && m_cnx.Connected)
			{
				m_cnx.Shutdown(SocketShutdown.Both);
				m_cnx.Close();
			}
			m_server.DropConnection(this);
		}
}

	// Incoming connections service provider
	public abstract class ServiceProvider: ICloneable
	{
		// Provides a new instance of the object.
		public virtual object Clone()
		{
			throw new Exception("Must be implemented by derived classes.");
		}

		// Gets executed when the server accepts a new connection.
		public abstract void OnAcceptConnection(ConnectionState state);

		// Gets executed when the server detects incoming data.
		// This method is called only if OnAcceptConnection has already finished.
		public abstract void OnReceiveData(ConnectionState state);

		// Gets executed when the server needs to shutdown the connection.
		public abstract void OnDropConnection(ConnectionState state);
	}

	// TCP server implementation
	public class Server
	{
		private int m_port;
		private Socket m_listener;
		private ServiceProvider m_provider;
		private ArrayList m_connections;
		private int m_maxConnections;

		private AsyncCallback OnConnectionReady;
		private WaitCallback OnAcceptConnection;
		private AsyncCallback OnReceivedDataReady;

		// Initializes server. To start accepting connections call Start method.
		public Server(ServiceProvider in_provider, int in_port, int in_maxcnx)
		{
			m_provider = in_provider;
			m_port = in_port;
			m_maxConnections = in_maxcnx;
			m_listener = new Socket(	AddressFamily.InterNetwork,
										SocketType.Stream,
										ProtocolType.Tcp);
			m_connections = new ArrayList();
			OnConnectionReady = new AsyncCallback(OnConnectionReadyCallback);
			OnAcceptConnection = new WaitCallback(OnAcceptConnectionCallback);
			OnReceivedDataReady = new AsyncCallback(OnReceivedDataReadyCallback);
		}

		// Start accepting connections.
		// A false return value tell you that the port is not available.
		public bool Start(String in_host)
		{
			try
			{
				m_listener.Bind(new IPEndPoint(IPAddress.Parse(in_host), m_port));
				m_listener.Listen(100); // max number of connections allowed
				m_listener.BeginAccept(OnConnectionReady, null);
				return true;
			}
			catch
			{
				return false;
			}
		}

		// Callback function: A new connection is waiting.
		private void OnConnectionReadyCallback(IAsyncResult in_ar)
		{
			lock (this)
			{
				if (m_listener == null) return;

				Socket conn = m_listener.EndAccept(in_ar);
				if (m_connections.Count >= m_maxConnections)
				{
					string msg = "sIBL_GUI_XSI_Server: Max number of connections reached.";
					conn.Send(Encoding.UTF8.GetBytes(msg), 0, msg.Length, SocketFlags.None);
					conn.Shutdown(SocketShutdown.Both);
					conn.Close();
				}
				else
				{
					//Start a new connection
					ConnectionState st = new ConnectionState();
					st.m_cnx = conn;
					st.m_server = this;
					st.m_provider = (ServiceProvider)m_provider.Clone();
					st.m_buffer = new byte[4];
					m_connections.Add(st);

					//Queue the rest of the job to be executed later
					ThreadPool.QueueUserWorkItem(OnAcceptConnection, st);
				}
				//Resume the listening callback loop
				m_listener.BeginAccept(OnConnectionReady, null);
			}
		}

		// Executes OnAcceptConnection method from the service provider.
		private void OnAcceptConnectionCallback(object in_state)
		{
			ConnectionState st = in_state as ConnectionState;
			try
			{
				st.m_provider.OnAcceptConnection(st);
			}
			catch
			{
				// report error to provider
			}

			//Starts the ReceiveData callback loop
			if (st.m_cnx.Connected)
			{
				st.m_cnx.BeginReceive(	st.m_buffer, 0, 0,
										SocketFlags.None,
										OnReceivedDataReady, st);
			}
		}

		// Executes OnReceiveData method from the service provider.
		private void OnReceivedDataReadyCallback(IAsyncResult in_ar)
		{
			ConnectionState st = in_ar.AsyncState as ConnectionState;
			st.m_cnx.EndReceive(in_ar);

			// Consider the following condition as a signal that the
			// remote host dropped the connection.
			if (st.m_cnx.Available == 0)
			{
				DropConnection(st);
			}
			else
			{
				try
				{
					st.m_provider.OnReceiveData(st);
				}
				catch
				{
					// report error to the provider
				}

				// Resume ReceivedData callback loop
				if (st.m_cnx.Connected)
				{
					st.m_cnx.BeginReceive(	st.m_buffer, 0, 0,
											SocketFlags.None,
											OnReceivedDataReady, st);
				}
			}
		}

		// Shutdown the server
		public void Stop()
		{
			lock (this)
			{
				m_listener.Close();
				m_listener = null;

				//Close all active connections
				foreach (object obj in m_connections)
				{
					ConnectionState st = obj as ConnectionState;
					try
					{
						st.m_provider.OnDropConnection(st);
					}
					catch
					{
						//some error in the provider
					}
					st.m_cnx.Shutdown(SocketShutdown.Both);
					st.m_cnx.Close();
				}
				m_connections.Clear();
			}
		}

		// Removes a connection from the list
		internal void DropConnection(ConnectionState in_st)
		{
			lock (this)
			{
				in_st.m_cnx.Shutdown(SocketShutdown.Both);
				in_st.m_cnx.Close();
				if (m_connections.Contains(in_st))
				{
					m_connections.Remove(in_st);
				}
			}
		}

		public int MaxConnections
		{
			get
			{
				return m_maxConnections;
			}
			set
			{
				m_maxConnections = value;
			}
		}

		public int CurrentConnections
		{
			get
			{
				lock (this) { return m_connections.Count; }
			}
		}
	}

} // XSI.TCP namespace

Addons/sIBL_GUI_XSI_Server/cssrc/sIBL_GUI_XSI_Server_Helpers.cs
using Softimage.XSIOM; // XSI object model
using Softimage.XSIMath;
using Softimage.XSIUtil;
using System;
using System.Collections.Generic;
using System.Text;

namespace XSI.Helpers
{
	public class Base
	{
		CXSIApplicationClass m_xsi;
		CXSIFactoryClass m_fact;
		CXSIUtilsClass m_utils;

		protected Base()
		{
			m_xsi = new CXSIApplicationClass();
			m_fact = new CXSIFactoryClass();
			m_utils = new CXSIUtilsClass();
		}
		protected bool Log(String str)
		{
			m_xsi.LogMessage(str, siSeverity.siVerbose);
			return true;
		}

		protected bool Info(String str)
		{
			m_xsi.LogMessage(str, siSeverity.siInfo);
			return true;
		}

		protected bool Error(String str)
		{
			m_xsi.LogMessage(str, siSeverity.siError);
			return true;
		}

		protected XSIApplication GetXSI()
		{
			return m_xsi;
		}

		protected XSIFactory GetFactory()
		{
			return m_fact;
		}

		protected XSIUtils GetUtils()
		{
			return m_utils;
		}
	}
}

Addons/sIBL_GUI_XSI_Server/cssrc/sIBL_GUI_XSI_Server_Service_Provider.cs
using System;
using System.Text;
using System.Collections;
using XSI.Helpers;

namespace XSI.TCP
{
	public class RequestsService: ServiceProvider
	{
		public override object Clone()
		{
			return new RequestsService();
		}

		public override void OnAcceptConnection(ConnectionState in_state)
		{
			if (!in_state.WriteTo(Encoding.UTF8.GetBytes("Connected to sIBL_GUI_XSI_Server!\r\n"), 0, 25))
			{
				in_state.EndConnection(); //if write fails... then close connection
			}
		}

		public override void OnReceiveData(ConnectionState in_state)
		{
			byte[] buffer = new byte[1024];
			while (in_state.AvailableData > 0)
			{
				int readBytes = in_state.ReadFrom(buffer, 0, 1024);
				if (readBytes > 0)
				{
					string receivedStr = Encoding.UTF8.GetString(buffer, 0, readBytes);

					// Put data on log stack
					ClientRequests requests = new ClientRequests();
					requests.AddEntry(receivedStr);
				}
				else
				{
					in_state.EndConnection(); //If read fails then close connection
				}
			}
		}

		public override void OnDropConnection(ConnectionState state)
		{
			//Nothing to clean here
		}
	}

	// Manages and process the requests received from clients
	class ClientRequests: Base
	{
		static public Queue m_requests = new Queue();

		public void AddEntry(String in_str)
		{
			lock (ClientRequests.m_requests.SyncRoot)
			{
				ClientRequests.m_requests.Enqueue(in_str);
			}
		}

		public void Log()
		{
			lock (ClientRequests.m_requests.SyncRoot)
			{
				foreach( String req in ClientRequests.m_requests)
				{
					Info(req);
				}
			}
		}

		public void Clear()
		{
			lock (ClientRequests.m_requests.SyncRoot)
			{
				ClientRequests.m_requests.Clear();
			}
		}

		public void Process()
		{
			lock (ClientRequests.m_requests.SyncRoot)
			{
				object param = null;
				while (ClientRequests.m_requests.Count > 0)
				{
					String req = (String)ClientRequests.m_requests.Dequeue();

                    Log("Executing Provided Script Request: " + req);
                    GetXSI().ExecuteScript( req, null, null, ref param);

				}
			}
		}
	}
}

Addons/sIBL_GUI_XSI_Server/Application/Plugins/sIBL_GUI_XSI_Server.py
"""
**sIBL_GUI_XSI_Server.py**

**Platform:**
	Windows, Linux.

**Description:**
	This module defines the sIBL_GUI_XSI_Server object.

**Others:**

"""

#**********************************************************************************************************************
#***	External imports.
#**********************************************************************************************************************
import SocketServer
import collections
import os
import re
import sys
import thread
from win32com.client import constants as siConstants

#**********************************************************************************************************************
#***	Internal imports.
#**********************************************************************************************************************
__sipath__ not in sys.path and sys.path.append(__sipath__)

from tcpServer import TCPServer

#**********************************************************************************************************************
#***	Module attributes.
#**********************************************************************************************************************
__author__ = "Thomas Mansencal"
__copyright__ = "Copyright (C) 2008 - 2012 - Thomas Mansencal"
__license__ = "GPL V3.0 - http://www.gnu.org/licenses/"
__maintainer__ = "Thomas Mansencal"
__email__ = "thomas.mansencal@gmail.com"
__status__ = "Production"

__all__ = []

#**********************************************************************************************************************
#***	Module classes and definitions.
#**********************************************************************************************************************
class LoggingRequestHandler(SocketServer.BaseRequestHandler):
	def handle(self):
		while True:
			data = self.request.recv(1024)
			if len(data) == 0:
				break
			Application.LogMessage(data)
		return True

class StackDataRequestHandler(SocketServer.BaseRequestHandler):
	def handle(self):
		while True:
			data = self.request.recv(1024)
			if len(data) == 0:
				break
			RuntimeGlobals.requestsStack.append(data)
		return True

class Constants(object):
	namespace = "sIBL_GUI_XSI_Server"
	author = __author__
	email = __email__
	website = "http://www.thomasmansencal.com/"
	majorVersion = 1
	minorVersion = 0
	patchVersion = 0
	defaultAddress = "127.0.0.1"
	defaultPort = 12288
	requestsHandler = StackDataRequestHandler
	languages = ("VBScript", "JScript", "Python", "PythonScript", "PerlScript")

class RuntimeGlobals(object):
	server = None
	requestsStack = collections.deque()

def XSILoadPlugin(pluginRegistrar):
	pluginRegistrar.Author = Constants.author
	pluginRegistrar.Name = Constants.namespace
	pluginRegistrar.URL = Constants.website
	pluginRegistrar.Email = Constants.email
	pluginRegistrar.Major = Constants.majorVersion
	pluginRegistrar.Minor = Constants.minorVersion

	pluginRegistrar.RegisterCommand("sIBL_GUI_XSI_Server_start", "sIBL_GUI_XSI_Server_start")
	pluginRegistrar.RegisterCommand("sIBL_GUI_XSI_Server_stop", "sIBL_GUI_XSI_Server_stop")
	pluginRegistrar.RegisterEvent("sIBL_GUI_XSI_Server_startupEvent", siConstants.siOnStartup)	
	pluginRegistrar.RegisterTimerEvent("sIBL_GUI_XSI_Server_timerEvent", 250, 0)

	Application.LogMessage("'{0}' has been loaded!".format(pluginRegistrar.Name))
	return True
	
def XSIUnloadPlugin(pluginRegistrar):
	stopServer()
	Application.LogMessage("'{0}' has been unloaded!".format(pluginRegistrar.Name))
	return True

def sIBL_GUI_XSI_Server_start_Init(context):
	Application.LogMessage("{0} | 'sIBL_GUI_XSI_Server_start_Init' called!".format(
	Constants.namespace), siConstants.siVerbose)
	return True

def sIBL_GUI_XSI_Server_start_Execute():
	Application.LogMessage("{0} | 'sIBL_GUI_XSI_Server_start_Execute' called!".format(
	Constants.namespace), siConstants.siVerbose)
	return True

def sIBL_GUI_XSI_Server_stop_Init(context):
	Application.LogMessage("{0} | 'sIBL_GUI_XSI_Server_stop_Init' called!".format(
	Constants.namespace), siConstants.siVerbose)
	return True

def sIBL_GUI_XSI_Server_stop_Execute():
	Application.LogMessage("{0} | 'sIBL_GUI_XSI_Server_stop_Execute' called!".format(
	Constants.namespace), siConstants.siVerbose)
	stopServer()
	return True

def sIBL_GUI_XSI_Server_startupEvent_OnEvent(context):
	Application.LogMessage("{0} | 'sIBL_GUI_XSI_Server_startupEvent_OnEvent' called!".format(
	Constants.namespace), siConstants.siVerbose)
	startServer()
	return True

def sIBL_GUI_XSI_Server_timerEvent_OnEvent(context):
	# Application.LogMessage("{0} | 'sIBL_GUI_XSI_Server_timerEvent' called!".format(
	# Constants.namespace), siConstants.siVerbose)	
	processData()
	return False

def startServer():
	RuntimeGlobals.server = TCPServer(Constants.defaultAddress, Constants.defaultPort, Constants.requestsHandler)
	RuntimeGlobals.server.start()
	return True 

def stopServer():
	RuntimeGlobals.server and RuntimeGlobals.server.stop()
	return True 

def processData():
	while RuntimeGlobals.requestsStack:
		data = RuntimeGlobals.requestsStack.popleft().strip()
		if os.path.exists(data):
			value = Application.ExecuteScript(data)
			Application.LogMessage(value)
		else:
			for language in Constants.languages:
				match = re.match(r"\s*(?P<language>{0})\s*\|(?P<code>.*)".format(language), data)
				if match:
					value = Application.ExecuteScriptCode(match.group("code"), match.group("language"))
					Application.LogMessage(value)
					break
	return True
